File: backend/controllers/navbarController.js
const db = require("../config/db");

exports.getNavbarMenu = async (req, res) => {
    try {
        const [rows] = await db.query(`
            SELECT 
                r.id AS region_id,
                r.name AS region_name,
                r.display_name,
                sr.id AS subregion_id,
                sr.name AS subregion_name,
                l.id AS location_id,
                l.name AS location_name
            FROM regions r
            LEFT JOIN subregions sr ON sr.region_id = r.id
            LEFT JOIN locations l ON l.subregion_id = sr.id
            ORDER BY r.id, sr.id, l.id\
        `);

        const result = {
            domestic: [],
            international: [],
        };

        rows.forEach((row) => {
            const targetGroup = row.display_name?.startsWith("Tour Miền") ? result.domestic : result.international;

            let region = targetGroup.find((r) => r.id === row.region_id);
            if (!region) {
                region = {
                    id: row.region_id,
                    name: row.region_name,
                    displayName: row.display_name,
                    destinations: [],
                };
                targetGroup.push(region);
            }
            // Tìm group (subregion) trong region.destinations
            let subregion = region.destinations.find((d) => d.group === row.subregion_name);
            if (!subregion) {
                subregion = {
                    id: row.subregion_id,
                    group: row.subregion_name,
                    places: [],
                };
                region.destinations.push(subregion);
            }

            // Thêm location vào places
            if (row.location_name) {
                subregion.places.push({id: row.location_id, name: row.location_name});
            }
        });

        res.json(result);
    } catch (error) {
        console.error("❌ Lỗi khi lấy navbar:", error.message);
        res.status(500).json({message: "Lỗi server"});
    }
};

File: BookingTourAPI/BookingTourAPI/BookingTourAPI/Controllers/NavbarController.cs
using BookingTourAPI.Data;
using BookingTourAPI.DTOs;
using BookingTourAPI.Services;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;

namespace BookingTourAPI.Controllers
{
    [ApiController]
    [Route("navbar-menu")]
    public class NavbarController : ControllerBase
    {
        private readonly BookingTourContext _context;
        private readonly ICacheService _cacheService;

        public NavbarController(BookingTourContext context, ICacheService cacheService)
        {
            _context = context;
            _cacheService = cacheService;
        }

        [HttpGet]
        public async Task<IActionResult> GetNavbarMenu()
        {
            try
            {
                var cacheKey = "navbar:menu";
                var cachedMenu = await _cacheService.GetAsync<NavbarMenuDto>(cacheKey);
                if (cachedMenu != null)
                {
                    return Ok(cachedMenu);
                }
                // Lấy tất cả region kèm subregion và location qua Include (Left Join tự động)
                var regions = await _context.Regions
                    .Include(r => r.Subregions)
                        .ThenInclude(sr => sr.Locations)
                    .OrderBy(r => r.Id)
                    .ToListAsync();

                var result = new NavbarMenuDto();

                foreach (var region in regions)
                {
                    var regionDto = new NavbarRegionDto
                    {
                        Id = region.Id,
                        Name = region.Name,
                        DisplayName = region.DisplayName
                    };

                    foreach (var subregion in region.Subregions.OrderBy(sr => sr.Id))
                    {
                        var destinationDto = new NavbarDestinationDto
                        {
                            Id = subregion.Id,
                            Group = subregion.Name
                        };

                        foreach (var location in subregion.Locations.OrderBy(l => l.Id))
                        {
                            destinationDto.Places.Add(new NavbarPlaceDto
                            {
                                Id = location.Id,
                                Name = location.Name
                            });
                        }

                        regionDto.Destinations.Add(destinationDto);
                    }

                    // Phân loại vào menu Trong nước (Domestic) hoặc Nước ngoài (International)
                    // Dựa vào logic gốc: Name hoặc DisplayName chứa "Miền" hoặc bắt đầu bằng "Tour Miền" thì là trong nước
                    var isDomestic = (region.DisplayName != null && (region.DisplayName.StartsWith("Tour Miền") || region.DisplayName.Contains("Miền"))) 
                                     || (region.Name != null && region.Name.Contains("Miền"));

                    if (isDomestic)
                    {
                        result.Domestic.Add(regionDto);
                    }
                    else
                    {
                        result.International.Add(regionDto);
                    }
                }

                await _cacheService.SetAsync(cacheKey, result, TimeSpan.FromHours(1));
                return Ok(result);
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[NAVBAR ERROR] {ex}");
                return StatusCode(500, new { message = "Lỗi server khi lấy navbar", error = ex.Message, inner = ex.InnerException?.Message });
            }
        }
    }
}
